File: src/main/kotlin/com/linqingying/cangjie/psi/CjIsExpression.java
package com.linqingying.cangjie.psi;

import com.linqingying.cangjie.CjNodeTypes;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CjIsExpression extends CjExpressionImpl implements CjOperationExpression {

    public CjIsExpression(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, @Nullable D data) {
        return visitor.visitIsExpression(this, data);
    }

    @NotNull
    public CjExpression getLeftHandSide() {
        return findChildByClass(CjExpression.class);
    }

    @Nullable
    @IfNotParsed
    public CjTypeReference getTypeReference() {
        return findChildByType(CjNodeTypes.TYPE_REFERENCE);
    }

    @Override
    @NotNull
    public CjSimpleNameExpression getOperationReference() {
        return findChildByType(CjNodeTypes.OPERATION_REFERENCE);
    }



}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjIsExpression.kt
package com.linqingying.cangjie.psi

import com.linqingying.cangjie.CjNodeTypes
import com.intellij.lang.ASTNode

class CjIsExpression(node: ASTNode) : CjExpressionImpl(node), CjOperationExpression {
    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D?): R {
        return visitor.visitIsExpression(this, data)
    }

    val leftHandSide: CjExpression
        get() = findChildByClass(CjExpression::class.java)!!

    @get:IfNotParsed
    val typeReference: CjTypeReference?
        get() = findChildByType(CjNodeTypes.TYPE_REFERENCE)

    override val operationReference : CjSimpleNameExpression get()   {
        return findChildByType(CjNodeTypes.OPERATION_REFERENCE)!!
    }
}
